tests/Tokenizer/Analyzer/AttributeAnalyzerTest.php

```diff
@@ -129,4 +130,295 @@ public static function provideIsAttributeCases(): iterable
 
         yield [true, '<?php #[Foo("(")] class Bar {}'];
     }
+
+    /**
+     * @requires PHP 8.0
+     *
+     * @dataProvider provideGetAttributeDeclarationsCases
+     *
+     * @param array{0: string, 1: list<AttributeAnalysis>} $expected
+     */
+    public function testGetAttributeDeclarations(string $code, int $startIndex, array $expected): void
+    {
+        $tokens = Tokens::fromCode($code);
+        $attributeDeclarationAnalyses = AttributeAnalyzer::collectFor($tokens, $startIndex);
+
+        self::assertSame(
+            serialize($expected),
+            serialize($attributeDeclarationAnalyses),
+        );
+    }
+
+    /**
+     * @return iterable<array{0: string, 1: int, 2: list<AttributeAnalysis>}>
+     */
+    public static function provideGetAttributeDeclarationsCases(): iterable
+    {
+        yield [
+            '<?php
+            function foo() {}
+            function bar() {}
+            ',
+            11,
+            [],
+        ];
```

Comment:
Please add these test cases:

```suggestion
        ];

        yield [
            '<?php
            #[Foo]
            function foo() {}
            function bar() {}
            ',
            15,
            [],
        ];

        yield [
            '<?php
            #[Baz]
            class Foo {}
            class Bar {}
            ',
            13,
            [],
        ];
```

`collectFor()` should stop on:
- first non-attribute meaningful token
- PHP open tag
- maybe other?

In general, it must not collect attributes belonging to other item.

Reply:
I kind of understand what you're aiming at, but I have some concerns. Let's consider this example:

```php
class Foo
{
    private string $foo;

    public function bar(Baz $baz, Qux $qux)
    {
        $corge = 3;
    }
}
```

Now, suppose the current index is on `$qux`. What exactly is the item in this context? Is it the parameter, meaning I would need to stop after `Qux` (which also implies identifying that it is a parameter and not a variable), or is the item the entire function? Similarly, if the index is on `$corge`, what is the item then? And in cases of class properties and constants, e.g., if I'm on `$foo`, is the item the property or the class itself?

I'm sure it's doable, but it sounds like it could get pretty complex, so I'm just checking if I understood you correctly.

Reply:
```php
<?php

#[A0]
class Abc {}

#[A1]
class Foo
{
    #[A2]
    private string $foo;

    #[A3]
    public const BAR = 'bar';

    public function bar(#[A4] Baz $baz, #[A5] Qux $qux)
    {
        $corge = 3;
    }
}

#[A6]
#[A7]
class Xyz {}
```

Based on example above, in _ideal scenario_, this should act like this (consider we run `collectFor()` for ...):

- `Foo` must collect `A1`
- `$foo` must collect `A2`
- `BAR` must collect `A3`
- `$baz` must collect `A4`
- `$qux` must collect `A5`
- `Xyz` must collect `A6` and `A7`
- `$corge` must not collect attributes

I believe we should allow collecting attributes for limited set of tokens, for example (because I probably can't figure out complete list right now):
- `T_CLASS` (classy elements `A1`, `A6+A7`)
- `T_VARIABLE` (class properties `A2`, methods' arguments `A4` and `A5`)
- `T_CONST` (class constants `A3`)

Unfortunately it can get pretty complex, so there's simplified solution: allow `collectFor()` working only with `T_ATTRIBUTE` (`#[`), but ensure all adjacent attributes are always collected. This IMHO should work like:
- rewind to first attribute in a group of attributes (e.g. if we run `collectFor()` for `A7` it should rewind to `A6`)
- collect all attributes until first non-attribute meaningful token (ignoring comments and whitespace)
- probably should be good to store some kind of cache with attributes found, cache key should be index of first `#[` in a group. This would ensure that if there are multiple `#[`, then only rewinding would be done for every attribute, but collecting would be done only once for the first attribute in each group. Cache should invalidate if `$tokens` is changed

The second approach looks less cumbersome and more stable (would support new allowed attribute targets out of the box). Does it sound reasonable to you 🙂?

PS. If we choose second approach, then probably `collectFor()` should be renamed to `collect()`.

Reply:
@HypeMC did you have time to get familiar with my feedback?

Reply:
@Wirone Sorry for not answering you, I've been a little busy. My plan is to give it a try this weekend. I believe your last approach makes sense. Will let you know how it goes :smile:

Reply:
@Wirone Sorry it took so long, but I finally got around to this again. I've made an [initial implementation](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7395/commits/ea08959ff2c0e3572d7b04bcda69a1d1bb68f88b) of your proposal, please let me know if this is what you were aiming for.